best-practice: add java best practice (#1956)
pingcap/docs: reference/best-practices/java-app.md

- Database connection pool: To reduce the overhead of creating a connection each time it is requested, applications usually use a connection pool to cache and reuse connections. JDBC [DataSource](https://docs.oracle.com/javase/8/docs/api/javax/sql/DataSource.html) defines a connection pool API. You can choose from different open-source connection pool implementations as needed.
- Data access framework: Applications usually use a data access framework such as [MyBatis](https://mybatis.org/mybatis-3/index.html) and [Hibernate](https://hibernate.org/) to further simplify and manage the database access operations.
- Application implementation: The application logic controls when to send what commands to the database. Some applications use [Spring Transaction](https://docs.spring.io/spring/docs/4.2.x/spring-framework-reference/html/transaction.html) aspects to manage transactions' begin and commit logics.

![Java application components](/media/java-practice-1.png)

As shown above, a Java application may do the following things:

> - Implement the MySQL protocol via the JDBC API to interact with TiDB

Reviewer: Please add a period `.`  in each of the following three items.
```suggestion
- Implement the MySQL protocol via the JDBC API to interact with TiDB.
```

- JDBC API usage choice
- API Implementer's parameter configuration

### JDBC API

For JDBC API usage, see [JDBC official tutorial](https://docs.oracle.com/javase/tutorial/jdbc/). This section covers the usage of several important APIs.

#### Use Prepare API

For OLTP (Online Transactional Processing) scenarios, the SQL statements sent by the program to the database are several types that can be exhausted after removing parameter changes. Therefore, it is recommended to use [Prepared Statements](https://docs.oracle.com/javase/tutorial/jdbc/basics/prepared.html) instead of regular [execution from a text file](https://docs.oracle.com/javase/tutorial/jdbc/basics/processingsqlstatements.html#executing_queries) and reuse Prepared Statements to execute directly. This avoids the overhead of repeatedly parsing and generating SQL execution plans in TiDB.

At present, most upper-level frameworks call the Prepare API for SQL execution. If you use the JDBC API directly for development, pay attention to choosing the Prepare API.

> In addition, with the default implementation of MySQL Connector/J, only client-side statements are preprocessed, and the statements are sent to the server in a text file after `?` is replaced on the client. Therefore, in addition to using the Prepare API, you also need to configure `useServerPrepStmts = true` in JDBC connection parameters before you perform statement preprocessing on the TiDB server (the parameter configuration section is described in detail below).

Reviewer: What about adding an anchor to locate the exact section?

Author: Okay.

- Periodically check whether the connection is available using a separate thread
- Send a test query regularly to keep alive connections

Different connection pool implementations may support one or more of the above methods. You can check your connection pool documentation to find the corresponding configuration.

## Data access framework

Applications often use some kind of data access framework to simplify database access.

### MyBatis

[MyBatis](http://www.mybatis.org/mybatis-3/) is currently a popular Java data access framework. It is mainly used to manage SQL queries and complete the mapping of result sets and Java objects. MyBatis is highly compatible with TiDB. MyBatis rarely has problems based on its historical issues.

Here we mainly focus on the following configurations.

#### Mapper parameter

MyBatis Mapper supports two parameters:

- `select 1 from t where id = #{param1}` will be converted to `select 1 from t where id =?` as a Prepared Statement and be "prepared", and the actual parameter will be used for reuse. You can achieve the best performance when using this parameter with the previously mentioned Prepare connection parameters.

> - `select 1 from t where id = ${param2}` will be replaced with `select 1 from t where id = 1` as a text file and be executed. If this statement is "prepared" to different parameters, it may cause TiDB to cache a large number of Prepared Statements, and executing SQL operations this way has injection security risks.

Author: Please help double-check the translation: @lysu 
如果这条语句被 prepare 成了不同参数，可能会导致 TiDB 缓存大量的 prepare 语句，并且这种方式执行 SQL 有注入安全风险

Author: So updates it to: 

If this statement is replaced with different parameters and is executed, MyBatis will send different requests for "preparing" the statements to TiDB. This might cause TiDB to cache a large number of...?

> - `select 1 from t where id = ${param2}` will be replaced with `select 1 from t where id = 1` as a text file and be executed. If this statement is "prepared" to different parameters, it may cause TiDB to cache a large number of Prepared Statements, and executing SQL operations this way has injection security risks.

Author: ```suggestion
- `select 1 from t where id = ${param2}` will be replaced with `select 1 from t where id = 1` as a text file and be executed. If this statement is replaced with different parameters and is executed, MyBatis will send different requests for "preparing" the statements to TiDB. This might cause TiDB to cache a large number of Prepared Statements, and executing SQL operations this way has injection security risks.
```

To support the automatic rewriting of multiple `INSERT` statements into the form of `insert ... values(...), (...), ...`, in addition to configuring `rewriteBatchedStatements=true` in JDBC as mentioned before, MyBatis can also use dynamic SQL to semi-automatically generate batch inserts. Take the following mapper as an example:

```xml
<insert id="insertTestBatch" parameterType="java.util.List" fetchSize="1">
  insert into test
   (id, v1, v2)
  values
  <foreach item="item" index="index" collection="list" separator=",">
  (
   #{item.id}, #{item.v1}, #{item.v2}
  )
  </foreach>
  on duplicate key update v2 = v1 + values(v1)
</insert>
```

This mapper generates an `insert on duplicate key update` statement. The number of `(?,?,?)` following "values" is determined by the number of passed lists. Its final effect is similar to using `rewriteBatchStatements=true`, which also effectively reduces communication overhead between the client and TiDB.

As mentioned before, you also need to note that the Prepared Statements will not be cached after their maximum length exceeds the value of `prepStmtCacheSqlLimit`.

#### Streaming result

> The earlier section introduces how to stream read execution results in JDBC. In addition to the corresponding configurations of JDBC, if you want to read a super large result set in MyBatis, you also need to note that:

Reviewer: What about setting an anchor in `The earlier section` to the actual earlier section?

Author: Updated

- You can set `fetchSize` for a single SQL statement in the mapper configuration (see the previous code block). Its effect is equivalent to calling `setFetchSize` in JDBC.
- You can use the query interface with `ResultHandler` to avoid getting the entire result set at once.
- You can use the `Cursor` class for stream reading.

If you configure mappings using XML, you can stream read results by configuring `fetchSize="-2147483648"`(`Integer.MIN_VALUE`) in the mapping's `<select>` section.

```xml
<select id="getAll" resultMap="postResultMap" fetchSize="-2147483648">
  select * from post;
</select>
```

> If you configure mappings using code, you can use `@Options(fetchSize = Integer.MIN_VALUE)` and return `Cursor` so that the SQL results can be read in streaming.

Author: Please help double-check the translation: @lysu
而使用代码配置映射，则可以使用 `@Options(fetchSize = Integer.MIN_VALUE)` 并返回 `Cursor` 从而让 SQL 结果能被流式读取。

> If you configure mappings using code, you can use `@Options(fetchSize = Integer.MIN_VALUE)` and return `Cursor` so that the SQL results can be read in streaming.

Author: 原文是指返回 `Cursor`类型的结果？

> If you configure mappings using code, you can use `@Options(fetchSize = Integer.MIN_VALUE)` and return `Cursor` so that the SQL results can be read in streaming.

Author: ```suggestion
If you configure mappings using code, you can add the `@Options(fetchSize = Integer.MIN_VALUE)` annotation and keep the type of results as `Cursor` so that the SQL results can be read in streaming.
```

> If you configure mappings using code, you can use `@Options(fetchSize = Integer.MIN_VALUE)` and return `Cursor` so that the SQL results can be read in streaming.

Author: Updated. PTAL @lysu

> If you configure mappings using code, you can use `@Options(fetchSize = Integer.MIN_VALUE)` and return `Cursor` so that the SQL results can be read in streaming.

Reviewer 2: > 原文是指返回 `Cursor`类型的结果？@lysu

yes

```java
@Select("select * from post")
@Options(fetchSize = Integer.MIN_VALUE)
Cursor<Post> queryAllPost();
```

### `ExecutorType`

You can choose `ExecutorType` during `openSession`. MyBatis supports three types of executors:

- Simple: The Prepared Statements are called to JDBC for each execution (if the JDBC configuration item `cachePrepStmts` is enabled, repeated Prepared Statements will be reused)
- Reuse: The Prepared Statements are cached in `executor`, so that you can reduce duplicate calls for Prepared Statements without using the JDBC `cachePrepStmts`

> - Batch: Each update statement will only be executed with `ExecuteBatch()` called when `addBatch()` is used for a query or a commit. If `rewriteBatchStatements` is enabled in the JDBC layer, it will try to rewrite the statements. If not, the statements will be sent one by one.

Author: Please help double-check the translation: @lysu
Batch：每次更新只有在 `addBatch` 到 query 或 commit 时才会调用 `executeBatch` 执行，如果 JDBC 层开启了 `rewriteBatchStatements`，则会尝试改写，没有开启则会一条条发送

> - Batch: Each update statement will only be executed with `ExecuteBatch()` called when `addBatch()` is used for a query or a commit. If `rewriteBatchStatements` is enabled in the JDBC layer, it will try to rewrite the statements. If not, the statements will be sent one by one.

Author: ```suggestion
- Batch: Each update operation (`INSERT`/`DELETE`/`UPDATE`) will first be added to the batch, and will be executed until the transaction commits or a `SELECT` query is performed. If `rewriteBatchStatements` is enabled in the JDBC layer, it will try to rewrite the statements. If not, the statements will be sent one by one.
```

> - Batch: Each update statement will only be executed with `ExecuteBatch()` called when `addBatch()` is used for a query or a commit. If `rewriteBatchStatements` is enabled in the JDBC layer, it will try to rewrite the statements. If not, the statements will be sent one by one.

Author: Updated

> Usually, the default value of `ExecutorType` is `Simple`. You need to change `ExecutorType` when calling `openSession`. If it is batch execution, you might find that in a transaction the update or insert statements are executed pretty fast, but it is slower when reading data or committing the transaction. This is actually normal, so you need to note this when troubleshooting slow SQL queries.

Reviewer: ```suggestion
Usually, the default value of `ExecutorType` is `Simple`. You need to change `ExecutorType` when calling `openSession`. If it is the batch execution, you might find that in a transaction the `update` or `insert` statements are executed pretty fast, but it is slower when reading data or committing the transaction. This is actually normal, so you need to note this when troubleshooting slow SQL queries.
```

## Spring Transaction

In the real world, applications might use [Spring Transaction](https://docs.spring.io/spring/docs/4.2.x/spring-framework-reference/html/transaction.html) and AOP aspects to start and stop transactions.

> By adding the `@Transactional` annotation to the method definition, AOP starts the transaction before the method is called, and commit the transaction before the method returns the result. If your application has a similar need, you can find `@Transactional` in code to determine when the transaction is started and closed.

Reviewer: ```suggestion
By adding the `@Transactional` annotation to the method definition, AOP starts the transaction before the method is called, and commits the transaction before the method returns the result. If your application has a similar need, you can find `@Transactional` in code to determine when the transaction is started and closed.
```

Pay attention to a special case of embedding. If it occurs, Spring will behave differently based on the [Propagation](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Propagation.html) configuration. Because TiDB does not support savepoint, nested transactions are not supported yet.

## Misc

> 

Reviewer: Please add one or two sentences to briefly introduce the following section.

Author: Added

### Troubleshooting tools

> Using the powerful troubleshooting tools of JVM is recommended when an issue occurs in your Java application and you do not know the business logic. Here are a few common tools:

Reviewer: ```suggestion
Using the powerful troubleshooting tools of JVM is recommended when an issue occurs in your Java application and you do not know the application logic. Here are a few common tools:
```

#### jstack

[jstack](https://docs.oracle.com/javase/7/docs/technotes/tools/share/jstack.html) is similar to pprof/goroutine in Go, which can easily troubleshoot the process stuck issue.

> By executing `jstack pid`, you can output the IDs and stack information of all threads in the target process. There is only the Java stack in the default output. If you want to output the C++ stack in the JVM at the same time, add the `-m` option.

Reviewer: ```suggestion
By executing `jstack pid`, you can output the IDs and stack information of all threads in the target process. By default, only the Java stack is output. If you want to output the C++ stack in the JVM at the same time, add the `-m` option.
```

> By using jstack multiple times, you can easily locate the stuck issue (such as updating only through Mybatis BatchExecutor flush) or the deadlock issue (such as the failure to send an SQL statement because test programs are preempting a lock in the application)

Author: Please help double-check the translation: @lysu
通过多次 jstack 可以方便地发现卡死问题（比如：都通过 Mybatis BatchExecutor flush 调用 update）或死锁问题（比如：测试程序都在抢占应用中某把锁导致没发送 SQL）

> By using jstack multiple times, you can easily locate the stuck issue (such as updating only through Mybatis BatchExecutor flush) or the deadlock issue (such as the failure to send an SQL statement because test programs are preempting a lock in the application)

Reviewer 2: how about

```
By using jstack multiple times, you can easily locate the stuck issue (such slow query from application's view due to using Batch ExecutorType in Mybatis) or the application deadlock issue (such as application do not send any SQL statements because it is preempting a lock before sending)
```

Author: Updated.

> By using jstack multiple times, you can easily locate the stuck issue (such as updating only through Mybatis BatchExecutor flush) or the deadlock issue (such as the failure to send an SQL statement because test programs are preempting a lock in the application)

Author: ```suggestion
By using jstack multiple times, you can easily locate the stuck issue (for example, a slow query from application's view due to using Batch ExecutorType in Mybatis) or the application deadlock issue (for example, the application does not send any SQL statement because it is preempting a lock before sending it)
```

> By using jstack multiple times, you can easily locate the stuck issue (for example, a slow query from application's view due to using Batch ExecutorType in Mybatis) or the application deadlock issue (for example, the application does not send any SQL statement because it is preempting a lock before sending it)

Reviewer: ```suggestion
By using jstack multiple times, you can easily locate the stuck issue (for example, a slow query from application's view due to using Batch ExecutorType in Mybatis) or the application deadlock issue (for example, the application does not send any SQL statement because it is preempting a lock before sending it).
```

In addition, `top -p $ PID -H` or Java swiss knife are common methods to view the thread ID. Also, to locate the issue of "a thread occupies a lot of CPU resources and I don't know what it is executing", do the following steps:

> - Use `printf "%x\n" pid` to convert the thread ID to hexadecimal

Reviewer: ```suggestion
- Use `printf "%x\n" pid` to convert the thread ID to hexadecimal.
```

> - Go to the jstack output to find the stack information of the corresponding thread

Reviewer: ```suggestion
- Go to the jstack output to find the stack information of the corresponding thread.
```

Note that mat only handles reachable objects by default. If you want to troubleshoot young GC issues, you can adjust mat configuration to view unreachable objects. In addition, for investigating the memory allocation of young GC issues (or a large number of short-lived objects), using Java Flight Recorder is more convenient.

#### trace

Online applications usually do not support modifying the code, but it is often desired that dynamic instrumentation is performed in Java to locate issues. Therefore, using btrace or arthas trace is a good option. They can dynamically insert trace code without restarting the application process.

#### Flame graph

Obtaining flame graphs in Java applications is tedious. For details, see [Java Flame Graphs Introduction: Fire For Everyone!](http://psy-lob-saw.blogspot.com/2017/02/flamegraphs-intro-fire-for-everyone.html).

## Conclusion

Based on commonly used Java components that interact with databases, this document describes the common problems and solutions for developing Java applications with TiDB. TiDB is highly compatible with the MySQL protocol, so most of the best practices for MySQL-based Java applications also apply to TiDB.

> Join us at [ASK TUG](https://asktug.com/), and share with broad TiDB user group about the practical experience or problems you run into when you use TiDB with Java applications.

Reviewer 2: maybe we can delete asktug link for English user

Author: Right.

> Join us at [ASK TUG](https://asktug.com/), and share with broad TiDB user group about the practical experience or problems you run into when you use TiDB with Java applications.

Author: ```suggestion
Join us at [TiDB Community slack channel](https://tidbcommunity.slack.com/archives/CH7TTLL7P), and share with broad TiDB user group about your experience or problems when you develop Java applications with TiDB.
```

> Based on commonly used Java components that interact with databases, this document describes the common problems and solutions for developing Java applications with TiDB. TiDB is highly compatible with the MySQL protocol, so most of the best practices for MySQL-based Java applications also apply to TiDB.

Reviewer: ```suggestion
Based on commonly used Java components that interact with databases, this document describes the common problems in and solutions for developing Java applications with TiDB. TiDB is highly compatible with the MySQL protocol, so most of the best practices for MySQL-based Java applications also apply to TiDB.
```

> Join us at [TiDB Community slack channel](https://tidbcommunity.slack.com/archives/CH7TTLL7P), and share with broad TiDB user group about your experience or problems when you develop Java applications with TiDB.

Reviewer: ```suggestion
Join us at [TiDB Community slack channel](https://tidbcommunity.slack.com/archives/CH7TTLL7P), and share with a broad TiDB user group about your experience or problems when you develop Java applications with TiDB.
```

#### Use `StreamingResult` to get the execution result

In most scenarios, to improve execution efficiency, JDBC obtains query results in advance and save them in client memory by default. But when the query returns a super large result set, the client often wants the database server to reduce the number of records returned at a time, and waits until the client's memory is ready and it requests for the next batch.

Usually, there are two kinds of processing methods in JDBC:

- [Set `FetchSize` to `Integer.MIN_VALUE`](https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-reference-implementation-notes.html#ResultSet) to ensure that the client does not cache. The client will read the execution result from the network connection through `StreamingResult`.
- To use Cursor Fetch, first [set `FetchSize`](http://makejavafaster.blogspot.com/2015/06/jdbc-fetch-size-performance.html) as a positive integer and configure `useCursorFetch=true` in the JDBC URL.

TiDB supports both methods, but it is preferred that you use the first method, because it is a simpler implementation and has a better execution efficiency.

### MySQL JDBC parameters

> JDBC usually provides implementation-related configurations in the form of JDBC URL parameters. This section introduces [MySQL Connector/J's parameter configurations](https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-reference-configuration-properties.html). (If you use MariaDB, see [MariaDB's parameter configurations](https://mariadb.com/kb/en/library/about-mariadb-connector-j/#optional-url-parameters)). Because this document cannot cover all configuration items, it mainly focuses on several parameters that might affect performance.

Reviewer: ```suggestion
JDBC usually provides implementation-related configurations in the form of JDBC URL parameters. This section introduces [MySQL Connector/J's parameter configurations](https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-reference-configuration-properties.html) (If you use MariaDB, see [MariaDB's parameter configurations](https://mariadb.com/kb/en/library/about-mariadb-connector-j/#optional-url-parameters)). Because this document cannot cover all configuration items, it mainly focuses on several parameters that might affect performance.
```
